src/payment/strategies/masterpass/masterpass-payment-strategy.ts

```diff
@@ -0,0 +1,153 @@
+    private _paymentMethod?: PaymentMethod;
+    private _walletButton?: HTMLElement;
+    private _walletCallback?: MasterpassCheckoutCallback;
+
+    constructor(
+        store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _masterpassScriptLoader: MasterpassScriptLoader
+    ) {
+        super(store);
+    }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId } = options;
+
+        this._paymentMethod = this._store.getState().paymentMethods.getPaymentMethod(methodId);
+
+        if (!this._paymentMethod) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        if (!options.masterpass) {
+            throw new InvalidArgumentError('Unable to initialize payment because "options.masterpass" argument is not provided.');
+        }
+
+        const walletButton = options.masterpass.walletButton && document.getElementById(options.masterpass.walletButton);
+
+        if (!this._hasPaymentInfo()) {
+            return this._masterpassClientSetup()
+                .then(checkoutCallback => {
+                    this._walletCallback = checkoutCallback;
+
+                    if (walletButton) {
+                        this._walletButton = walletButton;
+                        this._walletButton.addEventListener('click', this._handleWalletButtonClick);
+                    }
+                })
+                .then(() => super.initialize(options));
+        }
+
+        return super.initialize(options);
+    }
+
+    deinitialize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        this._paymentMethod = undefined;
+
+        if (this._walletButton) {
+            this._walletButton.removeEventListener('click', this._handleWalletButtonClick);
+        }
+
+        this._walletButton = undefined;
+        this._walletCallback = undefined;
+
+        return super.deinitialize(options);
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const order = { useStoreCredit: payload.useStoreCredit };
+
+        if (!this._paymentMethod) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const gateway = this._paymentMethod.initializationData.gateway;
+        const paymentData = this._paymentMethod.initializationData.paymentData;
+
+        if (!gateway) {
+            throw new InvalidArgumentError('Unable to proceed because "paymentMethod.initializationData.gateway" argument is not provided.');
+        }
+
+        if (!paymentData) {
+            throw new InvalidArgumentError('Unable to proceed because "paymentMethod.initializationData.paymentData" argument is not provided.');
```

Comment:
You need to show the wallet here instead of throwing an error. And then finalise the order upon redirection. Nevertheless, I'm happy for you to leave a comment here and we can address this later `TODO: Redirect to Masterpass if nonce has not been generated yet. And then finalise the order when the shopper is redirected back to the checkout page.`

Reply:
@michxff Let's add this TODO comment in.

Reply:
@Sure-Round added